In [1]:
import spacy

In [2]:
%matplotlib inline

import numpy as np
import matplotlib
import matplotlib.pyplot as plt
import jsonpickle
import json
import random
import tweepy
import spacy
import time
from datetime import datetime

In [3]:
class Tweets:
    
    
    def __init__(self,term="",corpus_size=100):
        self.tweets={}
        if term !="":
            self.searchTwitter(term,corpus_size)
                
    def searchTwitter(self,term,corpus_size):
        searchTime=datetime.now()
        while (self.countTweets() < corpus_size):
            new_tweets = api.search(term,lang="en",count=10)
            for nt_json in new_tweets:
                nt = nt_json._json
                if self.getTweet(nt['id_str']) is None and self.countTweets() < corpus_size:
                    self.addTweet(nt,searchTime,term)
            time.sleep(5)
                
    def addTweet(self,tweet,searchTime,term="",count=0):
        id = tweet['id_str']
        if id not in self.tweets.keys():
            self.tweets[id]={}
            self.tweets[id]['tweet']=tweet
            self.tweets[id]['count']=0
            self.tweets[id]['searchTime']=searchTime
            self.tweets[id]['searchTerm']=term
        self.tweets[id]['count'] = self.tweets[id]['count'] +1
        
    def getTweet(self,id):
        if id in self.tweets:
            return self.tweets[id]['tweet']
        else:
            return None
    
    def getTweetCount(self,id):
        return self.tweets[id]['count']
    
    def countTweets(self):
        return len(self.tweets)
    
    # return a sorted list of tupes of the form (id,count), with the occurrence counts sorted in decreasing order
    def mostFrequent(self):
        ps = []
        for t,entry in self.tweets.items():
            count = entry['count']
            ps.append((t,count))  
        ps.sort(key=lambda x: x[1],reverse=True)
        return ps
    
    # reeturns tweet IDs as a set
    def getIds(self):
        return set(self.tweets.keys())
    
    # save the tweets to a file
    def saveTweets(self,filename):
        json_data =jsonpickle.encode(self.tweets)
        with open(filename,'w') as f:
            json.dump(json_data,f)
    
    # read the tweets from a file 
    def readTweets(self,filename):
        with open(filename,'r') as f:
            json_data = json.load(f)
            incontents = jsonpickle.decode(json_data)   
            self.tweets=incontents
        
    def getSearchTerm(self,id):
        return self.tweets[id]['searchTerm']
    
    def getSearchTime(self,id):
        return self.tweets[id]['searchTime']
    
    def getText(self,id):
        tweet = self.getTweet(id)
        text=tweet['full_text']
        if 'retweeted_status'in tweet:
            original = tweet['retweeted_status']
            text=original['full_text']
        return text
                
    def addCode(self,id,code):
        tweet=self.getTweet(id)
        if 'codes' not in tweet:
            tweet['codes']=set()
        tweet['codes'].add(code)
        
   
    def addCodes(self,id,codes):
        for code in codes:
            self.addCode(id,code)
        
 
    def getCodes(self,id):
        tweet=self.getTweet(id)
        return tweet['codes']
    
    # NEW -ROUTINE TO GET PROFILE
    def getCodeProfile(self):
        summary={}
        for id in self.tweets.keys():
            tweet=self.getTweet(id)
            if 'codes' in tweet:
                for code in tweet['codes']:
                    if code not in summary:
                            summary[code] =0
                    summary[code]=summary[code]+1
        sortedsummary = sorted(summary.items(),key=operator.itemgetter(0),reverse=True)
        return sortedsummary

In [4]:
smoking=Tweets()
smoking.readTweets("tweets-smoking.json")
smoking.countTweets()

100

In [5]:
vaping=Tweets()
vaping.readTweets("tweets-vaping.json")
vaping.countTweets()

100

ok. so this is a bit opaque.  is sci-kit any clearer?

Trying the scikit spacy combination here ---

https://nicschrading.com/project/Intro-to-NLP-with-spaCy/

In [6]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.base import TransformerMixin
from sklearn.pipeline import Pipeline
from sklearn.svm import LinearSVC
from sklearn.feature_extraction.stop_words import ENGLISH_STOP_WORDS
from sklearn.metrics import accuracy_score
import string
import re

In [7]:
train = ["I love space. Space is great.", "Planets are cool. I am glad they exist in space", 
        "lol @twitterdude that is gr8", "twitter &amp; reddit are fun.", 
        "Mars is a planet. It is red.", "@Microsoft: y u skip windows 9?", 
        "Rockets launch from Earth and go to other planets.", "twitter social media &gt; &lt;", 
        "@someguy @somegirl @twitter #hashtag", "Orbiting the sun is a little blue-green planet."]
labelsTrain = ["space", "space", "twitter", "twitter", "space", "twitter", "space", "twitter", "twitter", "space"]

test = ["i h8 riting comprehensibly #skoolsux", "planets and stars and rockets and stuff"]
labelsTest = ["twitter", "space"]

In [8]:
vectorizer = CountVectorizer()

In [9]:
vectorizer.fit(train)

CountVectorizer(analyzer='word', binary=False, decode_error='strict',
        dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
        lowercase=True, max_df=1.0, max_features=None, min_df=1,
        ngram_range=(1, 1), preprocessor=None, stop_words=None,
        strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
        tokenizer=None, vocabulary=None)

In [10]:
print(vectorizer.vocabulary_)

{'love': 23, 'space': 39, 'is': 18, 'great': 13, 'planets': 31, 'are': 3, 'cool': 5, 'am': 0, 'glad': 10, 'they': 43, 'exist': 7, 'in': 17, 'lol': 22, 'twitterdude': 46, 'that': 41, 'gr8': 12, 'twitter': 45, 'amp': 1, 'reddit': 33, 'fun': 9, 'mars': 25, 'planet': 30, 'it': 19, 'red': 32, 'microsoft': 27, 'skip': 35, 'windows': 47, 'rockets': 34, 'launch': 20, 'from': 8, 'earth': 6, 'and': 2, 'go': 11, 'to': 44, 'other': 29, 'social': 36, 'media': 26, 'gt': 15, 'lt': 24, 'someguy': 38, 'somegirl': 37, 'hashtag': 16, 'orbiting': 28, 'the': 42, 'sun': 40, 'little': 21, 'blue': 4, 'green': 14}


In [11]:
vectorizer.get_feature_names()

['am',
 'amp',
 'and',
 'are',
 'blue',
 'cool',
 'earth',
 'exist',
 'from',
 'fun',
 'glad',
 'go',
 'gr8',
 'great',
 'green',
 'gt',
 'hashtag',
 'in',
 'is',
 'it',
 'launch',
 'little',
 'lol',
 'love',
 'lt',
 'mars',
 'media',
 'microsoft',
 'orbiting',
 'other',
 'planet',
 'planets',
 'red',
 'reddit',
 'rockets',
 'skip',
 'social',
 'somegirl',
 'someguy',
 'space',
 'sun',
 'that',
 'the',
 'they',
 'to',
 'twitter',
 'twitterdude',
 'windows']

In [12]:
vectorizer.vocabulary_.get("windows")

47

In [13]:
vector=vectorizer.transform(train)

In [14]:
print(vector.shape)
print(type(vector))
print(vector.toarray())

(10, 48)
<class 'scipy.sparse.csr.csr_matrix'>
[[0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0
  0 0 0 2 0 0 0 0 0 0 0 0]
 [1 0 0 1 0 1 0 1 0 0 1 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0
  0 0 0 1 0 0 0 1 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0
  0 0 0 0 0 1 0 0 0 0 1 0]
 [0 1 0 1 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0
  0 0 0 0 0 0 0 0 0 1 0 0]
 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 2 1 0 0 0 0 0 1 0 0 0 0 1 0 1 0 0 0
  0 0 0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1
  0 0 0 0 0 0 0 0 0 0 0 1]
 [0 0 1 0 0 0 1 0 1 0 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 1 0 1 0 0 1 0
  0 0 0 0 0 0 0 0 1 0 0 0]
 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0
  1 0 0 0 0 0 0 0 0 1 0 0]
 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  0 1 1 0 0 0 0 0 0 1 0 0]
 [0 0 0 0 1 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 0 

In [15]:
len(train)

10

https://machinelearningmastery.com/prepare-text-data-machine-learning-scikit-learn/

In [16]:
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer = TfidfVectorizer()
# tokenize and build vocab
vectorizer.fit(train)
# summarize
print(vectorizer.vocabulary_)
print(vectorizer.idf_)

{'love': 23, 'space': 39, 'is': 18, 'great': 13, 'planets': 31, 'are': 3, 'cool': 5, 'am': 0, 'glad': 10, 'they': 43, 'exist': 7, 'in': 17, 'lol': 22, 'twitterdude': 46, 'that': 41, 'gr8': 12, 'twitter': 45, 'amp': 1, 'reddit': 33, 'fun': 9, 'mars': 25, 'planet': 30, 'it': 19, 'red': 32, 'microsoft': 27, 'skip': 35, 'windows': 47, 'rockets': 34, 'launch': 20, 'from': 8, 'earth': 6, 'and': 2, 'go': 11, 'to': 44, 'other': 29, 'social': 36, 'media': 26, 'gt': 15, 'lt': 24, 'someguy': 38, 'somegirl': 37, 'hashtag': 16, 'orbiting': 28, 'the': 42, 'sun': 40, 'little': 21, 'blue': 4, 'green': 14}
[2.70474809 2.70474809 2.70474809 2.29928298 2.70474809 2.70474809
 2.70474809 2.70474809 2.70474809 2.70474809 2.70474809 2.70474809
 2.70474809 2.70474809 2.70474809 2.70474809 2.70474809 2.70474809
 1.78845736 2.70474809 2.70474809 2.70474809 2.70474809 2.70474809
 2.70474809 2.70474809 2.70474809 2.70474809 2.70474809 2.70474809
 2.29928298 2.29928298 2.70474809 2.70474809 2.70474809 2.70474809
 

In [17]:
vector = vectorizer.transform(train)
# summarize encoded vector
print(vector.shape)
print(vector.toarray())

(10, 48)
[[0.         0.         0.         0.         0.         0.
  0.         0.         0.         0.         0.         0.
  0.         0.43323568 0.         0.         0.         0.
  0.28646791 0.         0.         0.         0.         0.43323568
  0.         0.         0.         0.         0.         0.
  0.         0.         0.         0.         0.         0.
  0.         0.         0.         0.73657982 0.         0.
  0.         0.         0.         0.         0.         0.        ]
 [0.34989927 0.         0.         0.29744635 0.         0.34989927
  0.         0.34989927 0.         0.         0.34989927 0.
  0.         0.         0.         0.         0.         0.34989927
  0.         0.         0.         0.         0.         0.
  0.         0.         0.         0.         0.         0.
  0.         0.29744635 0.         0.         0.         0.
  0.         0.         0.         0.29744635 0.         0.
  0.         0.34989927 0.         0.         0.         0

ok. now -- how can I train on this? 
and how can I pass in my spacy tokens?

# plan

1. Go back to https://nicschrading.com/project/Intro-to-NLP-with-spaCy/

1.1 create NLP as per part 3 with hashtag tokenizer.

1.2. wrap that in a tokenizer that calls the spacy NLP parser and tokenizes with stops, etc.

1.3 use that tokenizer in a tf-df

1.4 follow that up with classifier. 

1.5 test


In [18]:
nlp = spacy.load('en')
def hashtag_pipe(doc):
    merged_hashtag = True
    while merged_hashtag == True:
        merged_hashtag = False
        for token_index,token in enumerate(doc):
            if token.text == '#':
                try:
                    nbor = token.nbor()
                    start_index = token.idx
                    end_index = start_index + len(token.nbor().text) + 1
                    if doc.merge(start_index, end_index) is not None:
                        merged_hashtag = True
                        break
                except:
                    pass
    return doc
nlp.add_pipe(hashtag_pipe,first=True)

In [19]:
def includeToken(tok):
    val =False
    if tok.is_stop == False:
        if tok.is_alpha == True: 
            if tok.text =='RT':
                val = False
            elif tok.pos_=='NOUN' or tok.pos_=='PROPN' or tok.pos_=='VERB':
                val = True
        elif tok.text[0]=='#' or tok.text[0]=='@':
            val = True
    if val== True:
        stripped =tok.lemma_.lower().strip()
        if len(stripped) ==0:
            val = False
        else:
            val = stripped
    return val

In [20]:
def filterTweetTokens(tokens):
    filtered=[]
    for t in tokens:
        inc = includeToken(t)
        if inc != False:
            filtered.append(inc)
    return filtered

In [21]:
def tokenizeText(sample):
    tokens = nlp(sample)
    filtered=[]
    for t in tokens:
        inc = includeToken(t)
        if inc != False:
            filtered.append(inc)
    return filtered 

1.3 - set up tokenizer in a vectorizer.

In [22]:
vectorizer = TfidfVectorizer(tokenizer=tokenizeText)
clf = LinearSVC()
pipe = Pipeline([('vectorizer', vectorizer), ('clf', clf)])

1.4 now I would do vectorizer. fit - need good input and labels..

In [23]:
smoking

In [24]:
vaping

## ok. to do this I think the following needs to happen
1. flatten the lists into pairs of text and category
2. shuffle and pick 0.8 of each contaat

In [29]:
def flattenTweets(tweets):
    flat=[]
    for i in tweets.getIds():
        text = tweets.getText(i)
        cat = tweets.getSearchTerm(i) 
        pair =(text,cat)
        flat.append(pair)
    return flat

In [78]:
def getTestTrainSplit(pairs,splitFactor=0.8):
    random.shuffle(pairs)
    split=int(len(pairs)*splitFactor)
    train=pairs[:split]
    test =pairs[split:]
    return train,test

In [89]:
def getTestTrain(smoking,vaping):
    fsmoke=flattenTweets(smoking)
    fvape=flattenTweets(vaping)
    strain,stest=getTestTrainSplit(fsmoke)
    vtrain,vtest=getTestTrainSplit(fvape)
    train=strain+vtrain
    test=stest+vtest
    random.shuffle(train)
    random.shuffle(test)
    return train,test

In [90]:
# ok. split established, but something is not right. need to ensure right split. 
# an then will need to split train and test into cateogires and texts

In [91]:
train,test=getTestTrain(smoking,vaping)

In [92]:
len(train)

160

In [97]:
len(test)

40

In [102]:
trainTweets,trainCats=zip(*train)

In [100]:
len(trainTweets)

160

In [101]:
trainTweets[0]

'Hemp 400mg blueberry hemp vaping oil drops tincture https://t.co/sflyEay7us  via @pinterest https://t.co/l3k4rVUpc8'

In [103]:
len(trainCats)

160

In [104]:
trainCats[:10]

('vaping',
 'smoking',
 'smoking',
 'smoking',
 'vaping',
 'smoking',
 'smoking',
 'smoking',
 'vaping',
 'smoking')

now try to run through the pipe.

In [105]:
pipe.fit(trainTweets,trainCats)

Pipeline(memory=None,
     steps=[('vectorizer', TfidfVectorizer(analyzer='word', binary=False, decode_error='strict',
        dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
        lowercase=True, max_df=1.0, max_features=None, min_df=1,
        ngram_range=(1, 1), norm='l2', preprocessor=None, smooth_idf=Tr...ax_iter=1000,
     multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
     verbose=0))])

 ok. did that do anything? Let's get the test

In [106]:
testTweets,testCats=zip(*test)

In [107]:
preds = pipe.predict(testTweets)

In [109]:
print("accuracy:", accuracy_score(testCats, preds))

accuracy: 0.975


ok. 97.5% accurate classification. can we look at the mismatches?

zip up input, category, prediction.  Look at first few

In [114]:
res=zip(testTweets,testCats,preds)
reslist=list(res)
reslist[:10]

[('#UNFAO is scaling up efforts on reducing the amount of #wood used as #fuel for #fish smoking in the #Gambia. With the new #UNFAO Thiaroye Technology stove, #women in the #fishsmoking &amp; drying #industry have improved access to #technology &amp; #livelihood. #ZeroHunger \n@FAOWestAfrica https://t.co/ifT8KSRo3O',
  'smoking',
  'smoking'),
 ('Lmao he smoking big bricks. 100? Bro gon have to miss me like aubrey https://t.co/b7SJrqsmH8',
  'smoking',
  'smoking'),
 ('Cochrane Podcast: Are there any smoking cessation programmes that can help adolescents to stop smoking?https://t.co/cyrRqGZmgV',
  'smoking',
  'smoking'),
 ('“I hate smoking my parents weed, it fucken sucks”', 'smoking', 'smoking'),
 ('me: smoking weed hasn’t affected me at all\n\nsomeone: count to 10\n\nme: https://t.co/SUoGzARpom',
  'smoking',
  'smoking'),
 ('I love smoking weed in beautiful ass places, looking at beautiful ass things.',
  'smoking',
  'smoking'),
 ('Limitless LMC Classic v2 220W Box Mod Preview #22

filter out those that don't match

In [116]:
res=zip(testTweets,testCats,preds)
for (sample,label,pred) in res:
    if label != pred:
        print(sample+", "+label+", "+pred)

Walked outside to avoid staffer wrath. Immediately took up smoking. Is there a difference between vaping and e-cigs?, vaping, smoking


ok. must put this into the text to challenge people to make the record better.  Then, challenge them to annotate and filter..